In [1]:
# Importing the necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

# Setting plots to be embeded in this notebook 
%matplotlib inline

# Setting the background style of seaborn
sb.set_style("darkgrid")

In [54]:
# Loading the dataset into a data frame that I call df_electricity
df_fuel = pd.read_csv("https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv")


In [55]:
# Peeping into the first five rows of the data frame
df_fuel.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [56]:
# Saving the data frame before continuation
df_fuel.to_csv("fuel_utility_data.csv", index = False)

In [57]:
# Viewing some samples of the data frame
df_fuel.sample(10, replace = False)

record_id  utility_id_ferc1  report_year  \
8049    f1_fuel_2000_12_138_1_1               138         2000   
12625   f1_fuel_2004_12_193_0_8               193         2004   
10674   f1_fuel_2002_12_134_2_1               134         2002   
22406    f1_fuel_2012_12_17_0_2                17         2012   
15590  f1_fuel_2006_12_159_2_12               159         2006   
27954    f1_fuel_2017_12_1_0_13                 1         2017   
21518   f1_fuel_2011_12_55_1_14                55         2011   
21157   f1_fuel_2011_12_44_2_10                44         2011   
16078    f1_fuel_2007_12_57_1_9                57         2007   
29088    f1_fuel_2018_12_44_1_8                44         2018   

         plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
8049              montour                coal       ton        1540190.0   
12625  presque isle-total                coal       ton        1810448.0   
10674          huntington                coal       ton        2696708.0   
22406           asheville                coal       ton         815705.0   
15590     burton combined                 oil       bbl            163.0   
27954  lawrenceburg - aeg                 gas       mcf        4007761.0   
21518              bartow                 gas       mcf         522109.0   
21157        st. clair pp                coal       ton        3519979.0   
16078         bowen no. 6                 oil       bbl           1441.0   
29088  trenton channel pp                 oil       bbl           9126.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
8049             25.338000                     35.260   
12625            21.028000                     31.335   
10674            22.642000                     17.923   
22406            25.356000                     92.867   
15590             5.796000                     40.633   
27954             1.073000                      3.563   
21518             1.019000                      3.079   
21157            19.116000                     39.945   
16078             5.817000                     90.174   
29088             5.832834                     87.094   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
8049                         33.320              1.39200  
12625                        31.335              1.49013  
10674                        17.533              0.79200  
22406                        92.566              3.66300  
15590                        78.790              7.14600  
27954                         3.551              3.32100  
21518                         5.774              3.02200  
21157                        45.791              2.09000  
16078                        90.174             15.31600  
29088                        90.476             14.93600

In [66]:
# Lowest fuel cost
df_fuel.groupby("fuel_type_code_pudl").mean()

utility_id_ferc1  report_year  fuel_qty_burned  \
fuel_type_code_pudl                                                   
coal                       115.637534  2004.623143     1.903473e+06   
gas                        124.122062  2005.978060     4.843183e+06   
nuclear                    107.397311  2002.970660     5.454838e+06   
oil                        113.236235  2007.195933     6.311677e+04   
other                      110.467066  2001.700599     2.948137e+06   
waste                      154.253968  2005.650794     1.171914e+05   

                     fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
fuel_type_code_pudl                                                   
coal                           20.762780                  67.421830   
gas                             1.021913                  13.659397   
nuclear                        17.582120                4955.157002   
oil                             5.814377                 168.877086   
other                           0.572752                  18.253856   
waste                           0.341163                  19.518122   

                     fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
fuel_type_code_pudl                                                     
coal                                   116.951141             1.787190  
gas                                     12.095172             5.408876  
nuclear                              28616.915039             1.716559  
oil                                    313.907691            12.698732  
other                                   16.871485           155.084910  
waste                                   18.413052           822.709937

> The fuel with lowest fuel cost per unit burned average is  $gas$. The average is 13.71 to 2 decimal places.

In [67]:
# Standard Deviation and 75th percentile of the measure of energy (Fuel_mmbtu_per_unit) per unit
df_fuel.fuel_mmbtu_per_unit.describe()

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

> For fuel_mmbtu_per_unit, the standard deviation is 10.53 and 75th percentile is 17.00.

In [68]:
# Skewness and Kurtosis  for fuel quantity burned
df_fuel.skew(axis = 0, skipna = True)

utility_id_ferc1                  0.605070
report_year                       0.006953
fuel_qty_burned                  15.851495
fuel_mmbtu_per_unit               4.135217
fuel_cost_per_unit_burned        19.787440
fuel_cost_per_unit_delivered    105.014227
fuel_cost_per_mmbtu             171.675535
dtype: float64

In [69]:
# Kurtosis for fuel_qty_burned
df_fuel.kurtosis(axis = 0, skipna = True)

utility_id_ferc1                    1.088432
report_year                        -1.145656
fuel_qty_burned                   651.369450
fuel_mmbtu_per_unit                55.595695
fuel_cost_per_unit_burned         485.255851
fuel_cost_per_unit_delivered    11765.054226
fuel_cost_per_mmbtu             29489.132594
dtype: float64

> The skewness and Kurtosis of fuel_qty_burned  are 15.85 and 651.37 respectively to 2 decimal places.

In [71]:
# Row with null, value, and percentage of all rows
df_fuel.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [72]:
# Getting all the rows
df_fuel.shape

(29523, 11)

In [93]:
# Getting the percentage
(180/df_fuel.shape[0])* 100

0.609694136774718

In [80]:
# Checking the data type to help decide null treatment
df_fuel.dtypes

record_id                        object
utility_id_ferc1                  int64
report_year                       int64
plant_name_ferc1                 object
fuel_type_code_pudl              object
fuel_unit                        object
fuel_qty_burned                 float64
fuel_mmbtu_per_unit             float64
fuel_cost_per_unit_burned       float64
fuel_cost_per_unit_delivered    float64
fuel_cost_per_mmbtu             float64
dtype: object

> The column with the highest number of nulls is fuel_unit. The number of nulls is 180. The percentage of nulls comapred to total number of rows is 0.61%.

> The column with nulls contains categorical variables. Therefore to replace the nulls we use the mode.

In [82]:
# Second and Third lowest correlation with fuel_cost_per_unit_burned
df_fuel.corrwith(df_fuel["fuel_cost_per_unit_burned"])

utility_id_ferc1               -0.037863
report_year                     0.013599
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_unit_burned       1.000000
fuel_cost_per_unit_delivered    0.011007
fuel_cost_per_mmbtu            -0.000437
dtype: float64

> The second and third lowest correlation with fuel_cost_per_unit_burned are fuel_mmbtu_per_unit      with  -0.010034 correlation coefficient and fuel_cost_per_unit_delivered with 0.011007 correlation coefficient  respectively. 

In [90]:
# Percentage Change in coal fuel type between 1994 and 1998
df_fuel.groupby(["fuel_type_code_pudl","report_year" ]).mean()

utility_id_ferc1  fuel_qty_burned  \
fuel_type_code_pudl report_year                                      
coal                1994               106.168421     1.509912e+06   
                    1995               107.854737     2.647153e+06   
                    1996               106.078886     1.621465e+06   
                    1997               106.124700     1.958999e+06   
                    1998               107.663573     2.084013e+06   
...                                           ...              ...   
waste               2014               157.777778     2.004032e+05   
                    2015               157.611111     2.076373e+05   
                    2016               149.823529     2.224943e+05   
                    2017               145.533333     2.356153e+05   
                    2018               142.312500     2.060233e+05   

                                 fuel_mmbtu_per_unit  \
fuel_type_code_pudl report_year                        
coal                1994                   21.670377   
                    1995                   21.604251   
                    1996                   21.507838   
                    1997                   21.582963   
                    1998                   21.375364   
...                                              ...   
waste               2014                    0.484179   
                    2015                    0.485471   
                    2016                    0.519342   
                    2017                    0.579088   
                    2018                    0.543370   

                                 fuel_cost_per_unit_burned  \
fuel_type_code_pudl report_year                              
coal                1994                         31.546467   
                    1995                         30.677442   
                    1996                         29.454299   
                    1997                        104.897309   
                    1998                         27.616234   
...                                                    ...   
waste               2014                         32.866333   
                    2015                         34.959444   
                    2016                         31.710529   
                    2017                         30.138600   
                    2018                         41.310750   

                                 fuel_cost_per_unit_delivered  \
fuel_type_code_pudl report_year                                 
coal                1994                           124.663939   
                    1995                            28.850834   
                    1996                            28.045432   
                    1997                            27.119043   
                    1998                           716.015659   
...                                                       ...   
waste               2014                            30.294056   
                    2015                            32.854444   
                    2016                            29.661471   
                    2017                            27.854533   
                    2018                            32.171813   

                                 fuel_cost_per_mmbtu  
fuel_type_code_pudl report_year                       
coal                1994                    1.436099  
                    1995                    1.397351  
                    1996                    1.349366  
                    1997                    1.322989  
                    1998                    1.292496  
...                                              ...  
waste               2014                   19.101611  
                    2015                   21.734111  
                    2016                   25.270529  
                    2017                23954.116867  
                    2018                   19.699562  

[148 rows x 6 columns]

> The fuel_cost_per_unit_burned in 1994 for $coal$ was 31.546467 and in 1998 it was 27.616234. The percentage difference is therefore calculated to be -12.46% to 2 decimal places.

In [92]:
# The year with the highest fuel cost per unit delivered
df_fuel.groupby("report_year").mean()

utility_id_ferc1  fuel_qty_burned  fuel_mmbtu_per_unit  \
report_year                                                           
1994               106.120648     2.724018e+06            10.638098   
1995               110.139051     3.712067e+06            10.644830   
1996               108.714154     2.544754e+06            10.778041   
1997               108.366545     2.813503e+06            10.247632   
1998               109.845528     3.267453e+06            10.704964   
1999               108.222857     3.281281e+06            10.558885   
2000               111.083758     2.472032e+06             9.224475   
2001               113.521386     2.321402e+06             8.371469   
2002               120.354357     2.145645e+06             8.304833   
2003               123.300578     1.888622e+06             8.403746   
2004               121.123322     1.782547e+06             8.332425   
2005               121.028369     1.768876e+06             8.135972   
2006               117.178600     1.957264e+06             8.138090   
2007               116.040348     2.035696e+06             8.027581   
2008               120.785016     2.046455e+06             7.892854   
2009               118.139116     2.154993e+06             7.688192   
2010               117.992863     2.423720e+06             7.575013   
2011               121.069355     2.461332e+06             7.786182   
2012               120.769912     2.747094e+06             7.796063   
2013               122.082569     2.741505e+06             7.797024   
2014               129.263877     2.700008e+06             7.588334   
2015               128.591949     3.340131e+06             7.217820   
2016               133.117021     3.693289e+06             6.846478   
2017               131.126888     3.352850e+06             6.781353   
2018               133.669821     4.178890e+06             6.596546   

             fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
report_year                                                            
1994                        213.191996                     63.636060   
1995                        354.422555                     32.735269   
1996                         79.434995                   9196.705948   
1997                        149.482132                  11140.197239   
1998                        126.683519                    287.154420   
1999                        120.053605                     25.551627   
2000                        237.797111                    985.362877   
2001                        133.013010                     60.050396   
2002                        101.179365                     47.594361   
2003                        123.183919                     55.663493   
2004                        154.897398                    139.524275   
2005                        107.111221                     41.438184   
2006                        151.361249                     38.657484   
2007                        202.622082                     43.325023   
2008                        255.954415                     58.588197   
2009                        215.699318                    652.694163   
2010                        262.951203                     91.862105   
2011                        301.754519                     59.774667   
2012                        258.804138                     60.994502   
2013                        255.991023                    172.307591   
2014                        442.359068                    192.737183   
2015                        237.933197                    326.535511   
2016                        247.083959                    103.901761   
2017                        246.281796                     46.196861   
2018                        232.992036                    499.269966   

             fuel_cost_per_mmbtu  
report_year                       
1994                    3.086977  
1995                    2.868838  
1996      

> The year with the highest fuel_cost_per_unit_delivered is 1997. This year has 11140.20 fuel_cost_per_unit_delivered to 2 decimal places